### PR TITLE
Implement WebSocket support for InvocationAgentServerHost

### PR DESCRIPTION
**Pull request by yulin-li**
# Implement WebSocket support for `InvocationAgentServerHost`

Adds the `invocations_ws` protocol — a long-lived, full-duplex WebSocket
transport — to `azure-ai-agentserver-invocations`, alongside the existing
`POST /invocations` HTTP endpoint.

## What's new

- **`@app.ws_handler` decorator** — registers an async function as the handler
  for `/invocations_ws`. The route is registered lazily on first decoration,
  so hosts without a registered handler return HTTP 404 to the upgrade.
- **Connection lifecycle** — the SDK calls `await websocket.accept()` before
  the handler runs, closes the socket cleanly on return (RFC 6455 code `1000`)
  and maps an unhandled handler exception to `1011`. Client-initiated
  `WebSocketDisconnect` codes are surfaced unchanged.
- **WebSocket Ping/Pong keep-alive** — disabled by default; enable by
  passing `InvocationAgentServerHost(ws_ping_interval=<seconds>)` or by
  setting the `WS_KEEPALIVE_INTERVAL` env var (auto-injected by AgentService
  into hosted-agent containers). Constructor arg takes precedence; `0`
  (or unset) disables. Wired to Hypercorn's `websocket_ping_interval` so
  Pings are sent at the WS protocol layer (opcode `0x9`) — survives Azure
  APIM / Azure Load Balancer's ~4-minute idle timeout without app-level
  heartbeats.
- **Telemetry** — structured close-event log line and OpenTelemetry span
  attributes `azure.ai.agentserver.invocations_ws.{session_id, close_code,
  duration_ms}`. Session ID honours `FOUNDRY_AGENT_SESSION_ID` so HTTP and
  WS sessions on the same container correlate.
- **Cancellation safety** — handler invocation wrapped in
  `try/except BaseException + finally` so `_finalize_session` always runs
  (emitting the close-event log line, ending the span, best-effort closing
  the socket) even when the task is cancelled, before the cancellation
  propagates.
- **Samples**
  - [`samples/ws_invoke_agent/`](sdk/agentserver/azure-ai-agentserver-invocations/samples/ws_invoke_agent/) — minimal echo over WS, plus the same host serving `POST /invocations`.
  - [`samples/ws_bidirectional_streaming_agent/`](sdk/agentserver/azure-ai-agentserver-invocations/samples/ws_bidirectional_streaming_agent/) — full-duplex agent: concurrent token streams + mid-flight cancel + `bye` control message.

## Implementation notes

- `_WSHandlerMixin` is a pure mixin (object subclass at runtime); typed as
  `AgentServerHost` only under `TYPE_CHECKING` so the diamond stays out of
  the runtime MRO.
- Signature validation at registration time (`async def` + invocable with one
  positional arg) — surfaces 0-arg / 2-required-arg mistakes at import,
  not at first request.
- `ws_handler` overwrite logs a warning to match the HTTP decorator's
  last-write-wins semantics without silently masking misconfiguration.

## Validation

- 188 tests (76 WS-specific) — `187 passed, 1 skipped`.
- Pylint, cspell, and the rest of the static gates pass.
- All 6 sample scripts compile.

## Follow-ups already merged into this branch

- #46885 — addresses PR review comments 18–52 (test coverage, sample
  polish, README close-codes table, OTel mock cleanup).
- Copilot AI review on this PR (6 threads, all replied + resolved):
  README sample-table wording, import order, two docstring fixes,
  cancellation finalize, sample `BaseException` narrowing.

## Checklist

- [x] CHANGELOG updated under `1.0.0b4 (Unreleased)`.
- [x] Tests cover decorator validation, route registration, lifecycle,
  close-codes, ping interval (arg / env / default), telemetry, multimodal
  payloads, edge cases.
- [x] Samples added with READMEs.
- [x] No breaking changes to existing `POST /invocations` API surface.

### CODE REVIEW

File: sdk/agentserver/azure-ai-agentserver-invocations/README.md [hunk 1]
```diff
@@ -196,6 +264,8 @@ Visit the [Samples](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/
 |---|---|
 | [simple_invoke_agent](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/agentserver/azure-ai-agentserver-invocations/samples/simple_invoke_agent/) | Minimal synchronous request-response |
 | [async_invoke_agent](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/agentserver/azure-ai-agentserver-invocations/samples/async_invoke_agent/) | Long-running operations with polling and cancellation |
+| [ws_invoke_agent](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/agentserver/azure-ai-agentserver-invocations/samples/ws_invoke_agent/) | Combined `POST /invocations` (HTTP) and `/invocations_ws` (WebSocket) host |
+| [ws_bidirectional_streaming_agent](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/agentserver/azure-ai-agentserver-invocations/samples/ws_bidirectional_streaming_agent/) | Full-duplex `/invocations_ws` agent: server-pushed heartbeats + concurrent token streams + mid-flight cancel |
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/README.md]: Thanks — fixed in d8a3... Updated the samples table row to drop 'server-pushed heartbeats' and clarify the sample relies on the SDK's WS Ping/Pong keep-alive instead.

File: sdk/agentserver/azure-ai-agentserver-invocations/tests/test_ws_multimodal_protocol.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Parity with :mod:`tests.test_multimodal_protocol` — covers binary frames,
+text frames (unicode + large), and JSON frames over WebSocket.
+"""
+from starlette.testclient import TestClient
+from starlette.websockets import WebSocket
+
+import pytest
+
+from azure.ai.agentserver.invocations import InvocationAgentServerHost
+
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/tests/test_ws_multimodal_protocol.py:14]: Thanks — fixed. Reordered imports so `pytest` is grouped with the third-party block (stdlib / third-party / local).

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+        :type websocket: ~starlette.websockets.WebSocket
+        :param session_id: Per-connection session ID for diagnostic logs.
+        :type session_id: str
+        :return: ``(close_code, exception_or_None)``.  ``close_code`` is the
+            RFC 6455 code that should be sent to the client; ``exception``
+            is set only for an *unhandled* exception (so the caller can map
+            it to span error events and a 1011 close).
+        :rtype: tuple[int, Optional[BaseException]]
+        :raises RuntimeError: If no handler is registered (programmer error;
+            the public ``_ws_endpoint`` checks ``_ws_fn`` before calling).
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py]: Thanks — fixed the docstring on `_invoke_user_handler` to say the route is only registered after `ws_handler` is decorated (so reaching the method without a handler is an SDK-internal bug, not a user misconfiguration). The behavior is unchanged.

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+            websocket=websocket,
+            span_ctx=span_ctx,
+            otel_span=otel_span,
+            session_id=session_id,
+            start_ns=start_ns,
+            close_code=close_code,
+            handler_exc=handler_exc,
+            error_code="internal_error" if handler_exc is not None else None,
+        )
+
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py]: Good catch — wrapped the `_invoke_user_handler` call in `try/except BaseException` + `finally` so `_finalize_session` always runs (emitting the close-event log line, ending the span, best-effort closing the socket) before the cancellation propagates.

File: sdk/agentserver/azure-ai-agentserver-invocations/samples/ws_bidirectional_streaming_agent/ws_bidirectional_streaming_agent.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        await websocket.send_json({"type": "done", "id": prompt_id})
+    except asyncio.CancelledError:
+        # Best-effort: tell the client we honoured their cancel.  Suppress
+        # any send error (the socket may already be closed) and re-raise
+        # so the caller observes the cancellation.  ``BaseException``
+        # rather than ``Exception`` so we don't accidentally swallow a
+        # nested cancellation while emitting the courtesy frame.
+        try:
+            await websocket.send_json({"type": "cancelled", "id": prompt_id})
+        except BaseException:  # pylint: disable=broad-exception-caught
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/samples/ws_bidirectional_streaming_agent/ws_bidirectional_streaming_agent.py]: Thanks — narrowed the courtesy-cancel `except` back to `Exception`. We deliberately let `KeyboardInterrupt`/`SystemExit` propagate; only ordinary send failures (closed socket) are swallowed.

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+                    await websocket.send_text(msg)
+
+        :param fn: An async function accepting a Starlette
+            :class:`~starlette.websockets.WebSocket` and returning ``None``.
+        :type fn: Callable[[WebSocket], Awaitable[None]]
+        :return: The original function (unmodified).
+        :rtype: Callable[[WebSocket], Awaitable[None]]
+        :raises TypeError: If *fn* is not an ``async def`` function, or its
+            signature is not callable with exactly one positional argument.
+        """
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:142]: Thanks — softened the docstring/error wording to 'invocable with a single positional argument' to match what `inspect.signature(fn).bind(None)` actually validates. Updated the regex in `test_ws_decorator_pattern.py` to match.

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_constants.py [hunk 7]
```diff
@@ -21,3 +21,34 @@ class InvocationConstants:
...
+    ROUTE_PATH = "/invocations_ws"
+
+    # Environment variable for platform-injected keep-alive override.
+    # AgentService injects this into every hosted-agent container so the
+    # platform can tune the WebSocket Ping cadence without redeploying
+    # the agent code.
+    # Resolution order: explicit ``ws_ping_interval=`` constructor arg →
+    # ``WS_KEEPALIVE_INTERVAL`` env var → disabled (``0``).
+    # ``0`` disables keep-alive.
+    ENV_WS_KEEPALIVE_INTERVAL = "WS_KEEPALIVE_INTERVAL"
```
Comment (ankitbko) [inline sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_constants.py]: I don't see this header in the spec - https://github.com/coreai-microsoft/foundrysdk_specs/blob/main/specs/agents/hosted_agents/container-spec/docs/container-image-spec.md

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation.py [hunk 8]
```diff
@@ -182,15 +202,19 @@ def __init__(
         self,
         *,
         openapi_spec: Optional[dict[str, Any]] = None,
+        ws_ping_interval: Optional[float] = None,
```
Comment (ankitbko) [inline sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation.py]: This should be part of AgentConfig https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/agentserver/azure-ai-agentserver-core/azure/ai/agentserver/core/_config.py

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation.py [hunk 9]
```diff
@@ -223,9 +247,45 @@ def __init__(
...
+                if self._ws_ping_interval == 0
+                else f"{self._ws_ping_interval}s"
+            ),
+        )
+
+    # ------------------------------------------------------------------
+    # Hypercorn server config (WebSocket Ping/Pong keep-alive)
+    # ------------------------------------------------------------------
+
+    def _build_hypercorn_config(self, host: str, port: int) -> object:
```
Comment (ankitbko) [inline sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation.py]: Lets just add it to the https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/agentserver/azure-ai-agentserver-core/azure/ai/agentserver/core/_base.py#L403

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+        start_ns = time.monotonic_ns()
+
+        # Open the OTel span before accepting so any tracecontext header
+        # the client sent is honoured for parenting child spans inside the
+        # user handler.  ``end_on_exit=False`` so we can attach the close
+        # code / duration before ending.
+        span_ctx = self.request_span(
+            websocket.headers,
+            session_id,
+            "websocket_session",
```
Comment (ankitbko) [inline sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:226]: How does websocket work with tracing? There should be a span `invoke_agent` somewhere for it to get integrated with tracing.

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+                handler_exc=exc,
+                error_code="accept_failed",
+            )
+            logger.error(
+                "WebSocket accept failed for session %s: %s",
+                session_id, exc, exc_info=True,
+            )
+            return
+
+        close_code: int = InvocationsWSConstants.CLOSE_NORMAL
```
Comment (ankitbko) [inline sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:258]: Are we propagating the otel context properly?

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+        session_id = self.config.session_id or str(uuid.uuid4())
+        start_ns = time.monotonic_ns()
+
+        # Open the OTel span before accepting so any tracecontext header
+        # the client sent is honoured for parenting child spans inside the
+        # user handler.  ``end_on_exit=False`` so we can attach the close
+        # code / duration before ending.
+        span_ctx = self.request_span(
+            websocket.headers,
+            session_id,
```
Comment (ankitbko) [inline sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:225]: What is invocation_id in websocket? How does "Get" work here? How can user cancel an invocation?

**Review by ankitbko — CHANGES_REQUESTED** [review]

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_constants.py (hunk 7 above)
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_constants.py]: will sync with you internally

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation.py (hunk 8 above)
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation.py]: thanks, moved

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation.py (hunk 9 above)
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation.py]: thanks, moved

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py (hunk 11 above)
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:258]: yes, this is just the default value, the following lines extract the close code correctly

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py (hunk 10 above)
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:226]: We landed on a **single connection-scoped span per WebSocket connection**: every WS upgrade opens one `websocket_session` OTel span (via `request_span(..., "websocket_session")`) that lives for the entire connection. It honours the client's `traceparent` header at handshake and carries the GenAI semantic-convention attributes plus `azure.ai.agentserver.invocations_ws.{session_id, close_code, duration_ms}`. Because `request_span` uses `start_as_current_span`, any child spans the handler opens (`tracer.start_as_current_span(...)`) are automatically parented to it via OTel's contextvars-based current-span tracking. We don't open a per-frame `invoke_agent` span: a WS connection is a single long-lived session, not a sequence of independently-identifiable HTTP invocations. README §"Per-connection tracing" documents this. Resolving — please re-open if you'd like a different model.

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py (hunk 12 above)
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:225]: There is **no per-message `invocation_id` for WebSocket** — a WS connection is one long-lived session, not a stream of independently-addressable HTTP invocations, so the HTTP-only `GET /invocations/{id}` and `POST /invocations/{id}/cancel` endpoints don't apply to WS sessions. Cancellation: clients cancel by closing the socket (or sending a close frame) — the server-side handler observes `asyncio.CancelledError` / `WebSocketDisconnect` and the SDK finalizes the span + emits the close-event log line with the appropriate code. README §"Per-connection tracing" documents this. Resolving.

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+
+Implements the ``@app.ws_handler`` decorator and the ``/invocations_ws``
+route described in the ``invocations_ws`` protocol spec.  The SDK wraps
+the user handler with:
+
+* ``await websocket.accept()`` before the handler runs;
+* WebSocket protocol-level Ping/Pong keep-alive (disabled by default;
+  enable via the ``WS_KEEPALIVE_INTERVAL`` env var or the
+  ``ws_ping_interval=`` constructor argument) so idle connections can
+  survive Azure APIM / Azure Load Balancer's ~4-minute idle timeout;
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py]: Fixed — the module docstring now only references the `WS_KEEPALIVE_INTERVAL` env var (commit a few minutes ago). The constructor kwarg path was removed when we moved `ws_ping_interval` to `AgentConfig`. Resolving.

File: sdk/agentserver/azure-ai-agentserver-invocations/samples/ws_bidirectional_streaming_agent/ws_bidirectional_streaming_agent.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+.. note::
+
+   Connection keep-alive is **not** an application concern: the SDK can
+   ask Hypercorn to send WebSocket protocol-level Ping frames
+   (opcode 0x9) every ``ws_ping_interval`` seconds (disabled by default;
+   enable by setting ``WS_KEEPALIVE_INTERVAL`` or by passing
+   ``InvocationAgentServerHost(ws_ping_interval=<seconds>)``).  When
+   enabled, that is enough to survive Azure APIM / Azure Load Balancer's
+   ~4-minute idle timeout without your handler having to push any
+   application-level heartbeat messages of its own.
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/samples/ws_bidirectional_streaming_agent/ws_bidirectional_streaming_agent.py]: Good catch — the bidirectional sample docstring had the same stale wording. Fixed: it now only mentions the `WS_KEEPALIVE_INTERVAL` env var (surfaced on `AgentConfig.ws_ping_interval`) and drops the `InvocationAgentServerHost(ws_ping_interval=<seconds>)` example. Resolving.

File: sdk/agentserver/azure-ai-agentserver-core/azure/ai/agentserver/core/_config.py [hunk 15]
```diff
@@ -123,6 +129,7 @@ def from_env(cls) -> Self:
                 _ENV_APPLICATIONINSIGHTS_CONNECTION_STRING, ""),
             otlp_endpoint=os.environ.get(_ENV_OTEL_EXPORTER_OTLP_ENDPOINT, ""),
             sse_keepalive_interval=resolve_sse_keepalive_interval(None),
+            ws_ping_interval=resolve_ws_ping_interval(None),
         )
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-core/azure/ai/agentserver/core/_config.py:133]: Agreed — dropped the unused `interval` kwarg from `resolve_ws_ping_interval` so the signature matches the actual (env-only) public API. The keep-alive interval is intentionally env-only (set by AgentService on the container); the helper now reads `WS_KEEPALIVE_INTERVAL` directly. Resolving.

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py (hunk 10 above)
Comment (ankitbko) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:226]: I may be wrong but I think `invoke_agent` is a reserved value for span that is used by system to identify LLM call.

**Review by ankitbko — APPROVED** [review]

File: sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py (hunk 10 above)
Comment (yulin-li) [reply sdk/agentserver/azure-ai-agentserver-invocations/azure/ai/agentserver/invocations/_invocation_ws.py:226]: only session level span is added here, because turn level / LLM call level is hard to detect as we don't know which returned messages are connected to one input message in websocket...

**Review by ankitbko — APPROVED** [review]
